server/src/test/java/org/elasticsearch/index/mapper/ObjectMapperMergeTests.java

```diff
@@ -198,6 +204,66 @@ public void testMergedFieldNamesMultiFieldsWithinSubobjectsFalse() {
+
+        // THEN "baz" new field is added to merged mapping
+        assertEquals(3, rootObjectMapper.mapperSize());
+        assertEquals(4, mergeWith.mapperSize());
+        assertEquals(3, mergedAdd0.mapperSize());
+        assertEquals(4, mergedAdd1.mapperSize());
+    }
+
+    public void testMergeWithLimitObjectField() {
+        RootObjectMapper root = new RootObjectMapper.Builder("_doc", Explicit.IMPLICIT_TRUE).build(MapperBuilderContext.root(false, false));
+        RootObjectMapper mergeWith = new RootObjectMapper.Builder("_doc", Explicit.IMPLICIT_TRUE).add(
+            new ObjectMapper.Builder("parent", Explicit.IMPLICIT_FALSE).add(
+                new KeywordFieldMapper.Builder("child1", IndexVersion.current())
+            ).add(new KeywordFieldMapper.Builder("child2", IndexVersion.current()))
+        ).build(MapperBuilderContext.root(false, false));
+
+        ObjectMapper mergedAdd1 = root.merge(mergeWith, MapperMergeContext.root(false, false, 1));
+        ObjectMapper mergedAdd2 = root.merge(mergeWith, MapperMergeContext.root(false, false, 2));
+        ObjectMapper mergedAdd3 = root.merge(mergeWith, MapperMergeContext.root(false, false, 3));
+        assertEquals(0, root.mapperSize());
+        assertEquals(1, mergedAdd1.mapperSize());
+        assertEquals(2, mergedAdd2.mapperSize());
+
+        ObjectMapper parent1 = (ObjectMapper) mergedAdd1.getMapper("parent");
+        assertNull(parent1.getMapper("child1"));
+        assertNull(parent1.getMapper("child2"));
+
+        ObjectMapper parent2 = (ObjectMapper) mergedAdd2.getMapper("parent");
+        // the order is not deterministic, but we expect one to be null and the other to be non-null
+        assertTrue(parent2.getMapper("child1") == null ^ parent2.getMapper("child2") == null);
+
+        ObjectMapper parent3 = (ObjectMapper) mergedAdd3.getMapper("parent");
+        assertNotNull(parent3.getMapper("child1"));
+        assertNotNull(parent3.getMapper("child2"));
+    }
+
+    public void testMergeWithLimitMultiField() {
+        RootObjectMapper mergeInto = new RootObjectMapper.Builder("_doc", Explicit.IMPLICIT_TRUE).add(
+            createTextKeywordMultiField("text", "keyword1")
+        ).build(MapperBuilderContext.root(false, false));
+        RootObjectMapper mergeWith = new RootObjectMapper.Builder("_doc", Explicit.IMPLICIT_TRUE).add(
+            createTextKeywordMultiField("text", "keyword2")
+        ).build(MapperBuilderContext.root(false, false));
+
+        assertEquals(2, mergeInto.mapperSize());
+        assertEquals(2, mergeWith.mapperSize());
+
+        ObjectMapper mergedAdd0 = mergeInto.merge(mergeWith, MapperMergeContext.root(false, false, 0));
+        ObjectMapper mergedAdd1 = mergeInto.merge(mergeWith, MapperMergeContext.root(false, false, 1));
+        assertEquals(2, mergedAdd0.mapperSize());
+        assertEquals(3, mergedAdd1.mapperSize());
+    }
```

Comment:
Should we expand these tests to cover for the following scenarios? 

- add new runtime fields - budget is taken into account
- add existing runtime fields - budget does not matter
- perhaps check that removing is not affected by the budget in any way, or maybe this is difficult to test, not sure
- adding the same object but with new sub-fields
- adding the same object with exactly the same sub-fields as already existing
- different combinations of subobjects and enabled perhaps

Reply:
makes sense, I'll look into it.

Reply:
I've added additional test cases.